### PR TITLE
TDL-17938 Updated bookmarking strategy

### PR DESCRIPTION
**Pull request by namrata270998**
# Description of change
- Updated the bookmarking strategy for advertisers, ads, adgroups and campaigns as it was returning None when no state was passed
- Updated the comparison strategy in ads, adgroups and campaings
- Added unittests

# Manual QA steps
 - Check that the tap works in case of no state is passed
 - Check that the streams ads, adgroups and campaings replicate properly when bookmark is passed.
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_tiktok_ads/streams.py [hunk 1]
```diff
@@ -145,7 +145,7 @@ def transform_ad_management_records(records, bookmark_value):
         # In case of an adgroup request, transform 'is_comment_disabled' type from integer to boolean
         if 'is_comment_disable' in record:
             record['is_comment_disable'] = bool(record['is_comment_disable'] == 0)
-        if bookmark_value is None or record['modify_time'] > bookmark_value:
+        if bookmark_value is None or utils.strptime_to_utc(record['modify_time']) > utils.strptime_to_utc(bookmark_value):
```
Comment (savan-chovatiya) [inline tap_tiktok_ads/streams.py:148]: Add PR comment here regarding the need for the above date transform.

File: tests/unittests/test_bookmark.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    def test_get_bookmark(self, test_transform_advertisers_records, test_get_bk_value):
+        '''
+        Verify that the get_bookmark() function returns {} when state for the stream is not passed.
+        '''
+        config = {"start_date": "test_start_date", "user_agent": "test_user_agent", "access_token": "test_at", "accounts": ['test_acc_id']}
+        state = {"bookmarks": {"campaigns": {"7052829480590606338": "2022-02-10T12:12:52.000000Z"}}}
+        advertisers = Advertisers(MockClient(), config, state)
+        stream =  MockCatalog('advertisers', 'advertisers', ['create_time'])
+        advertisers.process_batch(stream, [{'create_time': 1642114853}], 'test_acc_id')
+        test_get_bk_value.assert_called_with('advertisers', {}, 'test_acc_id')
```
Comment (savan-chovatiya) [inline tests/unittests/test_bookmark.py:43]: Add an inline comment here about which value is tested here from the argument.

File: tests/unittests/test_bookmark.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    def test_get_bookmark_value(self, test_transform_advertisers_records, test_pre_transform):
+        '''
+        Verify that the get_bookmark_value() function returns None when state for the Advertisers stream is not passed.
+        '''
+        config = {"start_date": "test_start_date", "user_agent": "test_user_agent", "access_token": "test_at", "accounts": ['test_acc_id']}
+        state = {"bookmarks": {"campaigns": {"7052829480590606338": "2022-02-10T12:12:52.000000Z"}}}
+        advertisers = Advertisers(MockClient(), config, state)
+        stream =  MockCatalog('advertisers', 'advertisers', ['create_time'])
+        advertisers.process_batch(stream, [{'create_time': 1642114853}], 'test_acc_id')
+        test_pre_transform.assert_called_with('advertisers', [{'create_time': 1642114853}], None)
```
Comment (savan-chovatiya) [inline tests/unittests/test_bookmark.py:56]: Add an inline comment here about which value is tested here from the argument.

File: tap_tiktok_ads/streams.py [hunk 4]
```diff
@@ -163,7 +163,10 @@ def transform_advertisers_records(records, bookmark_value):
 
 def get_bookmark_value(stream_name, bookmark_data, advertiser_id):
     if stream_name in ENDPOINT_ADVERTISERS:
-        return bookmark_data
+        if bookmark_data:
+            return bookmark_data
+        else:
+            return None
```
Comment (savan-chovatiya) [inline tap_tiktok_ads/streams.py:169]:
```suggestion
        if bookmark_data:
            return bookmark_data
        return None
```

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tap_tiktok_ads/streams.py (hunk 4 above)
Comment (namrata270998) [reply tap_tiktok_ads/streams.py]: Updated

File: tests/unittests/test_bookmark.py (hunk 3 above)
Comment (namrata270998) [reply tests/unittests/test_bookmark.py:60]: Added comment

File: tests/unittests/test_bookmark.py (hunk 2 above)
Comment (namrata270998) [reply tests/unittests/test_bookmark.py:46]: Added comment

File: tap_tiktok_ads/streams.py (hunk 1 above)
Comment (namrata270998) [reply tap_tiktok_ads/streams.py:150]: added comment

File: tap_tiktok_ads/streams.py [hunk 5]
```diff
@@ -163,7 +165,9 @@ def transform_advertisers_records(records, bookmark_value):
 
 def get_bookmark_value(stream_name, bookmark_data, advertiser_id):
```
Comment (dbshah1212) [inline tap_tiktok_ads/streams.py:166]: Do add function comment to explain the if condition.

Comment (namrata270998) [reply tap_tiktok_ads/streams.py:166]: Added comment

**Review by dbshah1212 — APPROVED** [review]

**Review by savan-chovatiya — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

**Pull request merged**